packages/doenetml-worker-rust/lib-doenetml-core/src/core/state_var_interfaces/text_state_var_interfaces.rs

```diff
@@ -0,0 +1,197 @@
+use crate::{
+    components::prelude::{
+        Dependency, DependencyInstruction, DependencyValueUpdateRequest,
+        RequestDependencyUpdateError, StateVarInterface, StateVarMutableViewTyped,
+        StateVarParameters, StateVarReadOnlyView, StateVarReadOnlyViewTyped,
+    },
+    dependency::DependencySource,
+    ExtendSource,
+};
+
+use super::common::create_dependency_instruction_from_extend_source;
+
+// use super::{
+//     StateVarInterface, StateVarMutableViewTyped, StateVarParameters, StateVarReadOnlyView,
+//     StateVarReadOnlyViewTyped,
+// };
+
+/// A string state variable interface that concatenates all string dependencies.
+///
+/// If `create_dependency_from_extend_source` is true and have an extend source extending this variable,
+/// then prepend the shadowed state variable.
+///
+/// If the state variable has a single dependency that is an essential state variable,
+/// then propagate the `used_default` attribute of the essential state variable.
+#[derive(Debug, Default)]
+pub struct GeneralStringStateVarInterface {
+    string_dependency_values: Vec<StateVarReadOnlyViewTyped<String>>,
+    from_single_essential: bool,
+}
+
+impl StateVarInterface<String> for GeneralStringStateVarInterface {
+    fn return_dependency_instructions(
+        &self,
+        extend_source: Option<&ExtendSource>,
+        parameters: &StateVarParameters,
+    ) -> Vec<DependencyInstruction> {
+        let mut dep_instructs: Vec<DependencyInstruction> = Vec::with_capacity(2);
+
+        if parameters.create_dependency_from_extend_source {
+            if let Some(dep_inst) =
+                create_dependency_instruction_from_extend_source(extend_source, parameters)
+            {
+                dep_instructs.push(dep_inst)
+            }
+        }
+
+        if let Some(dependency_instruction) = &parameters.dependency_instruction_hint {
+            dep_instructs.push(dependency_instruction.clone());
+        }
+
+        dep_instructs
+    }
+
+    fn save_dependencies_for_value_calculation(&mut self, dependencies: &Vec<Vec<Dependency>>) {
+        let num_dependencies = dependencies.iter().fold(0, |a, c| a + c.len());
+
+        let mut string_vals = Vec::with_capacity(num_dependencies);
+
+        for instruction in dependencies.iter() {
+            for Dependency {
+                value: dep_value, ..
+            } in instruction.iter()
+            {
+                if let StateVarReadOnlyView::String(dep_string_value) = dep_value {
+                    string_vals.push(dep_string_value.create_new_read_only_view())
+                } else {
+                    panic!("Got a non-string value for a dependency for a GeneralStringStateVarInterface");
+                }
+            }
+        }
```

Comment:
Would this look nicer with a `filter_map`?

Reply:
Assuming you mean `flat_map`, yes, I think it looks better especially with `.try_into`.